### PR TITLE
Esw/dynamixel control

### PR DESCRIPTION
**Pull request by hnsorens**
## Summary

- Added Dynamixel SDK as a ros2 package.
- Implemented a servo library.
- Changed mast_gimbal_hw_bridge to use servos instead of motors.
- Made a little change in the build system so that it would work on arch linux with conda (dont know if you want me to revert this or not).
- Added service for mast gimbal.

### Did you add documentation to the wiki?
No I have not.

## How was this code tested? 
The code was tested extensively on the dynamical servos, the soft limits were extensively unit tested to catch all edge cases.

### Did you test this in sim? 
No I did not, I don't think the servos are in the sim.

### Did you test this on the rover?
I did not test this on the rover, because at this time, the servos are not on the rover yet. I tested off of the rover.

### Did you add unit tests? 
I did not.

### CODE REVIEW

File: cmake/macros.cmake [hunk 1]
```diff
@@ -51,12 +51,10 @@ macro(mrover_add_component name sources includes)
...
-    install(CODE "execute_process( \
-    COMMAND ${CMAKE_COMMAND} -E create_symlink \
-    ${CMAKE_CURRENT_LIST_DIR}/../../build/${CMAKE_BUILD_TYPE}/${PROJECT_NAME}/lib${name}_component.so \
-    ${CMAKE_CURRENT_LIST_DIR}/../../install/${CMAKE_BUILD_TYPE}/${PROJECT_NAME}/lib/lib${name}_component.so \
-    )"
-)
+    install(
+        TARGETS ${name}_component
+        LIBRARY DESTINATION lib
+    )
```
Comment (edurso) [inline cmake/macros.cmake]: what libraries was this installing before? can you confirm that this still creates all the correct symlinks?

File: CMakeLists.txt [hunk 2]
```diff
@@ -315,6 +301,10 @@ if (NOT APPLE)
     mrover_add_header_only_library(can_device esw/can_device)
     target_link_libraries(can_device INTERFACE ${rclcpp_LIBRARIES} moteus)
 
+    mrover_add_library(servo_library esw/servo_library/*.cpp esw/servo_library)
+    ament_target_dependencies(servo_library rclcpp std_srvs dynamixel_sdk)
+    target_link_libraries(servo_library can_device units messaging parameter_utils dynamixel_sdk::dynamixel_sdk)
```
Comment (edurso) [inline CMakeLists.txt]: do not need to link this with `can_device` or `messaging`

File: CMakeLists.txt [hunk 3]
```diff
@@ -325,10 +315,12 @@ if (NOT APPLE)
         target_link_libraries(${name} can_device units messaging)
         target_link_libraries(${name} motor_library)
     endmacro()
-
+    
     mrover_add_esw_motor_bridge_node(drive_hw_bridge esw/drive_hw_bridge.cpp)
     mrover_add_esw_motor_bridge_node(arm_hw_bridge esw/arm_hw_bridge.cpp)
     mrover_add_esw_motor_bridge_node(mast_gimbal_hw_bridge esw/mast_gimbal_hw_bridge.cpp)
+    target_link_libraries(mast_gimbal_hw_bridge servo_library)
```
Comment (edurso) [inline CMakeLists.txt]: maybe a separate macro for servo nodes? or include in the motor node macro? maybe @jbrhm has some thoughts on that

File: .catkin_tools/profiles/profiles.yaml [hunk 4]
```diff
@@ -0,0 +1 @@
+active: default
```
Comment (jbrhm) [inline .catkin_tools/profiles/profiles.yaml]: Eric Mentioned this in his review.

File: esw/servo_library/servo.cpp [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+#define ADDR_CURRENT_LIMIT 102
+#define ADDR_VELOCITY_I_GAIN 76
+#define ADDR_VELOCITY_P_GAIN 78
+#define ADDR_POSITION_D_GAIN 80
+#define ADDR_POSITION_I_GAIN 82
+#define ADDR_POSITION_P_GAIN 84
+
+#define SERVO_POSITION_DEAD_ZONE 5
+
+#define SERVO_TICKS 4096
```
Comment (jbrhm) [inline esw/servo_library/servo.cpp]: nit. prefer `static constexpr`

File: esw/servo_library/servo.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+#include <dynamixel_sdk/packet_handler.h>
+#include <parameter.hpp>
+#include <string>
+
+#define DEFAULT_CURRENT_LIMIT 1750
+#define DEFAULT_POSITION_P_GAIN 400
+#define DEFAULT_POSITION_I_GAIN 0
+#define DEFAULT_POSITION_D_GAIN 0
+#define DEFAULT_VELOCITY_P_GAIN 180
+#define DEFAULT_VELOCITY_I_GAIN 0
```
Comment (jbrhm) [inline esw/servo_library/servo.hpp]: nit. prefer `static constexpr`

File: esw/servo_library/servo.cpp [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    position = (static_cast<float>(positionInt % 4096) / 4096.0f) * 360.0f;
+    return status;
+}
+
+auto Servo::getVelocity(ServoVelocity& velocity) -> Servo::ServoStatus
+{
+    uint8_t hardwareStatus;
+    uint32_t velocity_int;
+    Servo::ServoStatus status = read4Byte(ADDR_PRESENT_VELOCITY, velocity_int, &hardwareStatus);
+    velocity = (static_cast<float>(static_cast<int32_t>(velocity_int)) * 0.22888f);
```
Comment (jbrhm) [inline esw/servo_library/servo.cpp]: Adding a comment/constatnt for what 0.22888f represents would increase readability

**Comment by hnsorens** [comment]
Alright I think I got most of everything resolved. Let us know if something else needs to be fixed.

File: esw/mast_gimbal_hw_bridge.cpp [hunk 8]
```diff
@@ -26,80 +28,162 @@ namespace mrover {
...
-            mJointDataPub = create_publisher<sensor_msgs::msg::JointState>("mast_gimbal_joint_data", 1);
-            mControllerStatePub = create_publisher<msg::ControllerState>("mast_gimbal_controller_state", 1);
+        auto init() -> void {
 
-            mJointData.name = mMotorNames;
-            mJointData.position.resize(mMotorNames.size());
-            mJointData.velocity.resize(mMotorNames.size());
-            mJointData.effort.resize(mMotorNames.size());
+            
+            std::string device_name = this->declare_parameter<std::string>("device_name", "/dev/ttyUSB0");
```
Comment (jbrhm) [inline esw/mast_gimbal_hw_bridge.cpp]: Consider using `parameter.hpp`

**Review by jbrhm — APPROVED** [review]
lgtm 🔥
